streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -562,32 +561,41 @@ public void closeAndRecycleState() {
                 if (clean) {
                     stateMgr.flush();
                     recordCollector.flush();
-                    checkpoint = checkpointableOffsets();
+
+                    scheduleCheckpoint(checkpointableOffsets());
                 } else {
-                    checkpoint = null; // `null` indicates to not write a checkpoint
                     executeAndMaybeSwallow(false, stateMgr::flush, "state manager flush", log);
                 }
 
                 break;
 
             case RESTORING:
                 executeAndMaybeSwallow(clean, stateMgr::flush, "state manager flush", log);
-                checkpoint = Collections.emptyMap();
+                scheduleCheckpoint(emptyMap());
 
                 break;
 
             case SUSPENDED:
             case CLOSED:
                 // not need to checkpoint, since when suspending we've already committed the state
-                checkpoint = null; // `null` indicates to not write a checkpoint
-
                 break;
 
             default:
                 throw new IllegalStateException("Unknown state " + state() + " while prepare closing active task " + id);
         }
+    }
 
-        return checkpoint;
+    private void scheduleCheckpoint(final Map<TopicPartition, Long> checkpoint) {
+        this.checkpoint = checkpoint;
+        this.checkpointNeeded = true;
+    }
+
+    private void writeCheckpoint() {
```

Comment:
Maybe easier to make this `writeCheckpointIfNeeded`:
```
    private void writeCheckpointIfNeeded() {
        if (checkpointNeeded) {
            stateMgr.checkpoint(checkpoint);
            checkpointNeeded = false;
        }
    }
```

For this case, the caller does not need to check the `checkpointNeeded` flag?

Reply:
Sounds good